twilight-gateway/src/future.rs

```diff
@@ -0,0 +1,206 @@
+    message_future: Next<'a, Connection>,
+    /// Future resolving when the [`Shard`] must sent a heartbeat.
+    ///
+    /// [`Shard`]: crate::Shard
+    tick_heartbeat_future: TickHeartbeatFuture,
+}
+
+impl<'a> NextMessageFuture<'a> {
+    /// Initialize a new series of futures determining the next action to take.
+    pub fn new(
+        rx: &'a mut UnboundedReceiver<Message>,
+        message_future: Next<'a, Connection>,
+        maybe_heartbeat_interval: Option<Duration>,
+        maybe_last_sent: Option<Instant>,
+    ) -> Self {
+        Self {
+            channel_receive_future: rx,
+            message_future,
+            tick_heartbeat_future: TickHeartbeatFuture::new(
+                maybe_last_sent,
+                maybe_heartbeat_interval,
+            ),
+        }
+    }
+}
+
+impl Future for NextMessageFuture<'_> {
+    type Output = NextMessageFutureOutput;
+
+    fn poll(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        let mut this = self.as_mut();
+
+        if let Poll::Ready(()) = this.tick_heartbeat_future.poll_unpin(cx) {
+            return Poll::Ready(NextMessageFutureOutput::SendHeartbeat);
+        }
+
+        if let Poll::Ready(maybe_message) = this.channel_receive_future.poll_recv(cx) {
+            let message = maybe_message.expect("shard owns channel");
+
+            return Poll::Ready(NextMessageFutureOutput::UserChannelMessage(message));
+        }
+
+        if let Poll::Ready(maybe_try_message) = this.message_future.poll_unpin(cx) {
+            let maybe_message = maybe_try_message.and_then(Result::ok);
+
+            return Poll::Ready(NextMessageFutureOutput::Message(maybe_message));
+        }
+
+        Poll::Pending
+    }
+}
+
+/// Future that will resolve when the delay for a reconnect passes.
+///
+/// The duration of the future is defined by the number of attempts at
+/// reconnecting that have already been made. The math behind it is
+/// `2 ^ attempts`, maxing out at [`MAX_WAIT_SECONDS`].
+///
+/// [`MAX_WAIT_SECONDS`]: Self::MAX_WAIT_SECONDS
+pub struct ReconnectDelayFuture {
+    /// Inner future resolving when the duration passes.
+    inner: Pin<Box<Sleep>>,
+}
+
+impl ReconnectDelayFuture {
+    /// The maximum wait before resolving, in seconds.
+    const MAX_WAIT_SECONDS: u8 = 128;
+
+    /// Initialize a new unpolled future that will resolve when a reconnect
+    /// should be made.
+    pub fn new(reconnect_attempts: u8) -> Self {
+        let wait = 2_u8
+            .saturating_pow(u32::from(reconnect_attempts))
+            .min(Self::MAX_WAIT_SECONDS);
+
+        let duration = Duration::from_secs(u64::from(wait));
+
+        Self {
+            inner: Box::pin(time::sleep(duration)),
+        }
+    }
+}
+
+impl Future for ReconnectDelayFuture {
+    type Output = ();
+
+    fn poll(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        self.as_mut().inner.poll_unpin(cx)
+    }
+}
```

Comment:
Can just as easily be a function call
```suggestion
/// Future that will resolve when the delay for a reconnect passes.
///
/// The duration of the future is defined by the number of attempts at
/// reconnecting that have already been made. The math behind it is
/// `2 ^ attempts`, maxing out at `MAX_WAIT_SECONDS`.
pub async fn reconnect_delay(reconnect_attempts: u8) {
    /// The maximum wait before resolving, in seconds.
    const MAX_WAIT_SECONDS: u8 = 128;

    let wait = 2_u8
        .saturating_pow(reconnect_attempts.into())
        .min(MAX_WAIT_SECONDS);

    time::sleep(Duration::from_secs(wait.into())).await;
}
```